### ISSUE TITLE
__darwin_size_t is included even when it's not whitelisted

### ISSUE THREAD

**Issue by kornelski** [rust-lang-nursery/rust-bindgen#1215]
### Input C/C++ Header

```C++
#include <stdlib.h>
struct Foo {
    size_t s;
};
```

### Bindgen Invocation

```sh
$ bindgen --whitelist-type Foo test.h
```

on macOS High Sierra.

### Actual Results

```rust
pub type __darwin_size_t = :: std :: os :: raw :: c_ulong ; …
```

### Expected Results

No darwin internals at all.

**Comment by emilio** [comment]
By default a type generates all types that depend on it. In this case I guess in your platform is something like:

```c
typedef unsigned long long __darwing_size_t;
typedef __darwin_size_t size_t;
```

The special casing of size_t -> usize happens later in the pipeline (on code generation), and by then we think `__darwin_size_t` needs to be generated.

This one is fixable, not sure how easily in a clean way though, guess I'll need to give it a shot. Easy workaround would be to `blacklist_type("__darwin_.*")`.

**Comment by tmfink** [comment]
I'm hitting a similar issue on Linux. The `uint8_t` type is defined with two levels of typedef.

## Input C header
```c
// Emulate: #include <inttypes.h>
typedef unsigned char __uint8_t;
typedef __uint8_t uint8_t;

typedef uint8_t (*my_callback)(const uint8_t *code);
```

## Bindgen Invocation
```
$ bindgen extra_include.h --whitelist-type my_callback
```

## Actual Results

```rust
pub type __uint8_t = ::std::os::raw::c_uchar;
pub type my_callback = ::std::option::Option<unsafe extern "C" fn(code: *const u8) -> u8>;
```

## Expected Results

```rust
pub type my_callback = ::std::option::Option<unsafe extern "C" fn(code: *const u8) -> u8>;
```

Since `__uint8_t` is never used and not whitelisted, it should not appear in the output.

**Comment by kulp** [comment]
I think this is basically the same issue as #1188 -- an intermediate typedef is produced in both cases, because that is what the C headers say to do.

In my opinion, there should not be explicit encode edge cases in `bindgen` for this sort of thing.

The best description of rule to implement might be @tmfink's :
> Since `__uint8_t` is never used and not whitelisted, it should not appear in the output.

but I still am not sure why "a typedef is not used" should really mean that the typedef is suppressed *and other typedefs are stitched together in its absence*.

Since the C headers have this, why should not `bindgen` ? As @emilio says, `blacklist` is available.

**Comment by tmfink** [comment]
@kulp I see what you are saying about not wanting to special case this (assuming I'm understanding correctly).

One way to solve this would be to do an extra pass to see what items are "reachable" from the whitelisted set. Unreachable types (such as `__uint8_t` in my earlier example) would not be reachable and therefore removed.
If this incurs too much of a performance penalty, it could be opt-in.

My crate `capstone-sys` has pre-generate bindings that are used by default (to avoid cost of running bindgen each build).
https://github.com/capstone-rust/capstone-rs/blob/db25baef6381b339241302b3516792280dba5cc3/capstone-sys/pre_generated/capstone.rs#L3-L11

It would be nice if the bindgen output was consistent for each platform (Windows, Mac, Linux, FreeBSD, etc.) and did not rely details of the system headers. The blacklist approach depends on details of system's headers.

**Comment by kulp** [comment]
> One way to solve this would be to do an extra pass to see what items are "reachable" from the whitelisted set. Unreachable types (such as `__uint8_t` in my earlier example) would not be reachable and therefore removed.

I think actually the `__uint8_t` type *is* reachable from the whitelisted types (`uint8_t` -> `__uint8_t` -> `unsigned char`), but later during code generation the `uint8_t` type itself is changed to `u8` (as pointed out in https://github.com/rust-lang/rust-bindgen/issues/1215#issuecomment-355735454) so final bindings emit `__uint8_t` unnecessarily.

At the risk of overgeneralizing the problem, maybe @tmfink's request could be expressed like this :

_Prune type aliases (`typedef`s) that are not used by the emitted bindings._

However, this still does not really address what @kornelski wants, because the emitted bindings in that original case *do* reference the "internals" :
```rust
pub type __darwin_size_t = ::std::os::raw::c_ulong;
pub type size_t = __darwin_size_t;
#[repr(C)]
#[derive(Debug, Copy, Clone)]
pub struct Foo {
    pub s: size_t,
}
```

So, on my Mac at least, I do not see any ["special casing of `size_t` -> `usize`"](https://github.com/rust-lang/rust-bindgen/issues/1215#issuecomment-355735454) and I think there might be two separate issues here :
1. ignore "internal" intermediate typedefs (@kornelski) -- unclear how to determine what constitutes "internals"
1. prune unused typedefs (@tmfink) -- a clear rule can exist here

**Comment by emilio** [comment]
Yeah, this is related to `--size-t-is-usize` and co too. The issue is that we do the replacement for types like `uint8_t` to `u8` when we're generating code, which seems generally reasonable, but that means that we don't account for that replacement when computing the whitelist, and thus types that `uint8_t` uses are implicitly generated.

**Comment by emilio** [comment]
The right way to fix this is doing those replacements sooner, IMO, but that may or may not be a lot of work for this edge case.

**Comment by kornelski** [comment]
There are more options:

* Don't include intermediate typedefs or unused types from system headers. This would automatically exclude __darwin and similar OS/compiler implementation details.

* Eagerly resolve typedefs all the way to a primitive type. That's all what counts for an ABI.

* Always prune `__*` types, because that's C's way of saying "this is private".

**Comment by emilio** [comment]
Fair enough :)

Here some trade-offs with those that come to mind:

>     * Don't include intermediate typedefs or unused types from system headers. This would automatically exclude __darwin and similar OS/compiler implementation details.

Yeah, the problem is that this requires yet another pass on top of all the existing passes we do just to detect this special case, which is not great. You get this for free if you do the replacement earlier...

>     * Eagerly resolve typedefs all the way to a primitive type. That's all what counts for an ABI.

That's true, but it seems like the resulting bindings could be less idiomatic unnecessarily. So much stuff would end up being `void*` or `int` instead of some better name in the function signatures.

>     * Always prune `__*` types, because that's C's way of saying "this is private".

This is the simplest, but but if that typedef is used in a "public" struct, like (pseudo-code):

```c
struct something {
    // Public members, then...
    __uint32_t __reserved[10];
};
```

Then bindgen would generate bindings that don't compile, which is not great. Maybe we could auto-treat them as opaque and it'd "just work", not sure...

**Comment by kornelski** [comment]
To clarity, by eagerly resolve I've meant the intermediate types, so that you'd still get `type typedefed = u8` instead of `type typedefed = __internal_byte`.

And if a name is used, then of course it should not be pruned.

AFAIK "requires yet another pass" is purely bindgen's problem, not something user would be concerned about. C headers and libclang have messy issues, so needing multiple passes to clean that data up seems normal.
